text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Passing Dynamic Attributes/Properties
+status: DRAFTED
+created_at: 2021-07-05
+updated_at: YYYY-MM-DD
+champion: Mohammed Abdul Sattar (abdulsattar)
+pr: https://github.com/salesforce/lwc-rfcs/pull/52
+---
+
+# Passing Dynamic Attributes/Properties
+
+## Summary
+This proposal adds mechanism to pass a dynamic attributes or properties bag to elements in
+an LWC template.
+
+## Example
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+    childProps = { 
+        name: 'My Child',
+        style: 'width: 50px',
+        class: 'my-class',
+        onclick: this.handleClick,
+    };
+    
+    handleClick() {
+        console.log('Clicked')
+    }
+}
+```
+```html
+// x/myComponent.html
+<template>
+    <x-child lwc:spread={childProps}></x-child>
+</template>
+```
+
+The new `lwc:spread` directive passes `childProps` as properties/attributes to `x-child`.
+
+## Motivation
+
+In LWC, there is no way to pass props/attributes to a child component when the component author
+doesn't know in advance all the props/attributes that are needed.
+Adding this feature can be helpful in creating wrapper components that extend the functionality
+of an existing component while being oblivious to all the props that component requires.
+
+### Prior Art
+- React ([JSX Spread Operator](https://reactjs.org/docs/jsx-in-depth.html#spread-attributes))
+- Vue ([`v-bind`](https://v3.vuejs.org/api/directives.html#v-bind))
+- Svelte ([Spread Props](https://svelte.dev/tutorial/spread-props))
+- Lit ([Discussion for support](https://github.com/lit/lit/issues/923))
+
+## Detailed Design
+
+A new LWC directive, `lwc:spread` will be introduced that can be used to spread the props/attributes
+to the children elements.
+
+### Structure
+It accepts an object with keys as the attribute names and values as the
+corresponding attribute values.
+
+LWC Template compiler separates all attributes into:
+1. Attributes
+2. Properties
+3. `style` attribute
+4. `class` attribute
+
+The same algorithm will be used for separating those attributes at runtime.
+
+### Disallowed keys
+Some keys (e.g. 'for:each') alter the compiled HTML output itself, so they are not allowed.
+`lwc:spread` does not allow LWC directives in general.
```

Comment:
Is this valid?

```html
<template>
  <template for:each={many} for:item={one}>
    <c-child lwc:spread={one}></c-child>
  </template>
</template>
```

Reply:
Yes, I think the paragraph is confusing... but essentially it is saying that `{one}` object cannot have a key called `for:each`, because that's not a valid property name... I think that should be implicit if we say we only support props... directives are NOT props... but I think it is worth noticing, just make it more clear in the paragraph.